### PR TITLE
Bug fixes for DnD implementation in `libEGM`

### PR DESCRIPTION
**Pull request by dc03**
Rewrite parts of the DnD conversion code to more exactly mirror the java implementation in `EnigmaWriter.java` (https://github.com/enigma-dev/lgmplugin/blob/master/org/enigma/EnigmaWriter.java#L1173, which is the original LGM implementation)

Currently, the issue lies in object IDs not being properly generated which means that even though the code is generated more or less exactly as LGM does it, the objects being modified in the code are incorrect which can lead to a broken game.

### CODE REVIEW

File: CommandLine/libEGM/action.cpp [hunk 1]
```diff
@@ -113,13 +109,45 @@ std::string Actions2Code(const std::vector< buffers::resources::Action >& action
...
+            } else if (applies_to != "") {
+              code += std::string{"with ("} + ref_as_int(applies_to) + ") {";
+            } else {
+              code += "/*null with*/{";
+            }
+          }
+        }
 
         if (action.is_question()) {
-          code += "__if__ = ";
```
Comment (RobertBColton) [inline CommandLine/libEGM/action.cpp:122]: As discussed, you can't remove this fix that I added for if statement scoping.
Otherwise you'll break the scoping of variables inside if actions which can also have applies to.
The applies to/calling if inside a with statement is necessary so we don't have to prepend `self.` to all the variables in the if expression.

**Review by RobertBColton — CHANGES_REQUESTED** [review]
Keep looking for what's causing the 1945 bugs that are not in LGM and fix them.

File: CommandLine/libEGM/action.cpp (hunk 1 above)
Comment (dc03) [reply CommandLine/libEGM/action.cpp:122]: As per your suggestions, I have reverted the changes and added `__if__` back.

**Comment by dc03** [comment]
@RobertBColton I request you to look over this once again. In its current state, 1945v3 has somehow started working again, and it seems the only difference between what LGM and libEGM generate now is one `with()` statement and a `event_collision->name`, which I think is also caused by the `with()` having a different ID. The diff of the generated code is available here https://gist.github.com/dc03/dcc76bf54b50b3fe833919deb24db7b6, where LGM's output is the removed stuff and libEGM's output is the added stuff.

To get a useful output from `git diff`, I have `sort`ed the generated output because the C++ gets generated in different orders for some reason, which is why the diff output may look a little weird.

File: CommandLine/libEGM/action.cpp [hunk 2]
```diff
@@ -75,6 +75,39 @@ std::string Actions2Code(const std::vector< buffers::resources::Action >& action
...
+        } else {
+          return std::to_string(action.id());
+        }
+      };
+
+      if (action.is_question()) {
+        if (applies_to == "other") {
+          code += "with (other) ";
+        } else if (!applies_to.empty()) {
+          code += std::string{"with ("} + ref_as_int(applies_to) + ") ";
```
Comment (fundies) [inline CommandLine/libEGM/action.cpp]: "with ("s is shorter but "with (" should also work because ref_as_in is str. I don't think ref_as_int is needed at all though.

File: CommandLine/libEGM/action.cpp [hunk 3]
```diff
@@ -75,6 +75,39 @@ std::string Actions2Code(const std::vector< buffers::resources::Action >& action
   for (const auto &action : actions) {
     const auto &args = action.arguments();
 
+    std::string applies_to = action.who_name();
+    if (action.use_apply_to() && applies_to != "self") {
+      const auto &ref_as_int = [&action](const std::string &applies_to) -> std::string {
+        if (applies_to == "self") {
+          return "-1";
```
Comment (fundies) [inline CommandLine/libEGM/action.cpp]: Idont think self / other need to be converted to ints

File: CommandLine/libEGM/action.cpp (hunk 2 above)
Comment (dc03) [reply CommandLine/libEGM/action.cpp]: I have already gotten rid of this code, please check the latest commit :)

File: CommandLine/libEGM/action.cpp (hunk 3 above)
Comment (RobertBColton) [reply CommandLine/libEGM/action.cpp]: I mentioned that because the enum containing these keywords itself maps them to the ints.

**Review by RobertBColton — CHANGES_REQUESTED** [review]
It would be nice to see unit tests added for the drag and drop covering the weird edge cases.
Including things like applies to and the argument relative.

File: CommandLine/libEGM/action.cpp [hunk 4]
```diff
@@ -148,16 +149,19 @@ std::string Actions2Code(const std::vector< buffers::resources::Action >& action
...
 
-        if (action.relative())
+        if (action.use_relative())
           code += action.is_question() ? ");" : "\n}";
         if (action.is_question()) {
           code += "\nif (__if__)";
         }
+        code += "\n";
+
+        if (action.who_name() != "self" && !action.is_question())
```
Comment (RobertBColton) [inline CommandLine/libEGM/action.cpp:159]: This may work for GMX, but I don't know that it works for GMK because GMK only contains integers. For GMK it will be a `-1` instead of self. Would probably make sense to fix it in the GMK reader, in which case you can bring back that code you had in the lambda but invert the mapping it does.

Comment (dc03) [reply CommandLine/libEGM/action.cpp:159]: Wouldn't it be easier to have something like `(action.who_name() != "self" || action.id() != -1)` which will work for both cases?

Comment (RobertBColton) [reply CommandLine/libEGM/action.cpp:159]: It largely doesn't matter but I'm not sure this is the only place to handle it, that's why I think it's better to handle at the source.

**Comment by dc03** [comment]
@RobertBColton as per the discussions we had, I have redone the code generation for DnD a bit, and I have also added a stack to track the targets of actions. However, it can still be better, as targets are not tracked across non-nested, i.e. linear actions, so multiple with-statements can still be generated like
```gml
  with((obj_1))
    enigma::varaccess_test_x(int(self))+= 5;
  with((obj_1))
    enigma::varaccess_test_y(int(self))+= 1;
  with((obj_1))
  {
    __if__ = ...
    if((__if__))
      with((self))
    {
     ...
```

**Review by RobertBColton — APPROVED** [review]
The changes generally improve the behavior and clarity of generated drag and drop.

PS: I'm not merging as I do not understand the current state of the CI and why it is failing some of the style checks.

**Comment by dc03** [comment]
@RobertBColton you can basically ignore MINGW because of gRPC being broken upstream, which causes linker errors. For MSVC, I was fixing it in https://github.com/enigma-dev/RadialGM/pull/230  where the CI broke because of windows-latest upgrading to VS 2022. I do not know about any style checks, I didn't even know there were any

**Review by fundies — APPROVED** [review]
